In [16]:
with open('data/day_2.txt') as f:
    for line in f:
        program_str = line

print(program_str)

1,0,0,3,1,1,2,3,1,3,4,3,1,5,0,3,2,9,1,19,1,19,6,23,2,6,23,27,2,27,9,31,1,5,31,35,1,35,10,39,2,39,9,43,1,5,43,47,2,47,10,51,1,51,6,55,1,5,55,59,2,6,59,63,2,63,6,67,1,5,67,71,1,71,9,75,2,75,10,79,1,79,5,83,1,10,83,87,1,5,87,91,2,13,91,95,1,95,10,99,2,99,13,103,1,103,5,107,1,107,13,111,2,111,9,115,1,6,115,119,2,119,6,123,1,123,6,127,1,127,9,131,1,6,131,135,1,135,2,139,1,139,10,0,99,2,0,14,0


In [17]:
program_split = program_str.split(',')

program = [int(value) for value in program_split]

print(program)

[1, 0, 0, 3, 1, 1, 2, 3, 1, 3, 4, 3, 1, 5, 0, 3, 2, 9, 1, 19, 1, 19, 6, 23, 2, 6, 23, 27, 2, 27, 9, 31, 1, 5, 31, 35, 1, 35, 10, 39, 2, 39, 9, 43, 1, 5, 43, 47, 2, 47, 10, 51, 1, 51, 6, 55, 1, 5, 55, 59, 2, 6, 59, 63, 2, 63, 6, 67, 1, 5, 67, 71, 1, 71, 9, 75, 2, 75, 10, 79, 1, 79, 5, 83, 1, 10, 83, 87, 1, 5, 87, 91, 2, 13, 91, 95, 1, 95, 10, 99, 2, 99, 13, 103, 1, 103, 5, 107, 1, 107, 13, 111, 2, 111, 9, 115, 1, 6, 115, 119, 2, 119, 6, 123, 1, 123, 6, 127, 1, 127, 9, 131, 1, 6, 131, 135, 1, 135, 2, 139, 1, 139, 10, 0, 99, 2, 0, 14, 0]


Restore the program to its state before the crash:

In [18]:
program[1] = 12
program[2] = 2

print(program)

[1, 12, 2, 3, 1, 1, 2, 3, 1, 3, 4, 3, 1, 5, 0, 3, 2, 9, 1, 19, 1, 19, 6, 23, 2, 6, 23, 27, 2, 27, 9, 31, 1, 5, 31, 35, 1, 35, 10, 39, 2, 39, 9, 43, 1, 5, 43, 47, 2, 47, 10, 51, 1, 51, 6, 55, 1, 5, 55, 59, 2, 6, 59, 63, 2, 63, 6, 67, 1, 5, 67, 71, 1, 71, 9, 75, 2, 75, 10, 79, 1, 79, 5, 83, 1, 10, 83, 87, 1, 5, 87, 91, 2, 13, 91, 95, 1, 95, 10, 99, 2, 99, 13, 103, 1, 103, 5, 107, 1, 107, 13, 111, 2, 111, 9, 115, 1, 6, 115, 119, 2, 119, 6, 123, 1, 123, 6, 127, 1, 127, 9, 131, 1, 6, 131, 135, 1, 135, 2, 139, 1, 139, 10, 0, 99, 2, 0, 14, 0]


What value is left in position 0 after the program halts?

In [6]:
# Attempt 1: The last four positions tell us: opcode 2 means multiply the values in 
# positions 0 and 14, then store them in position 0.

print(program[0] * program[14])

0


This is incorrect. Probably need to run the program: 

In [102]:
def run_opcode(program):
    
    index = range(0, len(program))
    
    for i in index:
        if program[i] == 1:
            program[program[i + 3]] = program[i + 1] + program[i + 2]
        elif program[i] == 2:
            program[program[i + 3]] = program[i + 1] * program[i + 2]
        elif program[i] == 99:
            break
        else:
            continue
    
    return program

# TO DO: Make it skip ahead 4 after a 1 or 2

Run some tests:

In [103]:
test1 = [1, 0, 0, 0, 99]
print(run_opcode(test1))

[0, 0, 0, 0, 99]


Ah I see - the last 4 positions won't run because they're preceded by 99. Looking before, the last opcode was to add the values in position 139 and 10 and store them in position 0. 

In [105]:
test2 = [2,3,0,3,99]
print(run_opcode(test2))

[2, 3, 0, 0, 99]


556


In [70]:
test2 = [0, 1, 2]
for i in range(0, len(test2)):
    print(i)

0
1
2


In [93]:
test1 = [1, 0, 0, 0, 99]
for i in test1:
    print(i)

1
0
0
0
99
